[CORE-9126] `compression`: correct endianness in `snappy_java_compressor`

WillemKauf opened the pull request:
Issue: https://github.com/redpanda-data/redpanda/issues/25091

The versions in the `snappy` header are written using big-endian format in the `snappy-java` client used by kafka.

Mistakenly, `redpanda` would write them using little-endian format in our `snappy_java_compressor` implementation.

Correct this by encoding and decoding the `version` and `compatible_version` headers using big-endian format in `snappy_java_compressor`.

For references to `snappy-java`'s big-endian implementation, see:

* https://github.com/xerial/snappy-java/blob/65e1ec3de1a0d447b137c6dd6393629aa3d75b8b/src/main/java/org/xerial/snappy/SnappyOutputStream.java#L343-L349
* https://github.com/xerial/snappy-java/blob/65e1ec3de1a0d447b137c6dd6393629aa3d75b8b/src/main/java/org/xerial/snappy/SnappyCodec.java#L78-L81

## Backports Required

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v24.3.x
- [X] v24.2.x
- [ ] v24.1.x

## Release Notes

### Bug Fixes

* Fix the endianness of `snappy_java_compressor` headers to match that of `snappy-java`

WillemKauf committed:
`compression`: move `snappy_magic` to header

WillemKauf committed:
`compression`: correct endianness in `snappy_java_compressor`

The versions in the snappy header are written using big-endian format in the
`snappy-java` client used by kafka.

Mistakenly, `redpanda` would write them using little-endian format in our
`snappy_java_compressor` implementation.

Correct this by encoding and decoding the `version` and `compatible_version`
headers using big-endian format in `snappy_java_compressor`.

For references to `snappy-java`'s big-endian implementation, see:

* https://github.com/xerial/snappy-java/blob/65e1ec3de1a0d447b137c6dd6393629aa3d75b8b/src/main/java/org/xerial/snappy/SnappyOutputStream.java#L343-L349
* https://github.com/xerial/snappy-java/blob/65e1ec3de1a0d447b137c6dd6393629aa3d75b8b/src/main/java/org/xerial/snappy/SnappyCodec.java#L78-L81

WillemKauf committed:
`compression`: add `snappy_tests.cc`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--e0915a8c-629f-466e-a0a7-2c2d38e9aa85-->
<!--TEST RESULTS-->
<!--0691983b-e7e1-40bf-ae8e-800673940abf-->
<!--TEST RESULTS-->
<!--df8a9af6-558a-456d-b6e0-3c414d768979-->
<!--TEST RESULTS-->
<!--b96f93a1-00f0-43fb-ad00-42c8ac072feb-->
<!--TEST RESULTS-->
<!--ab43ac54-318e-48f1-b47c-110bb41b6ede-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61892">build#61892</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_gc_entire_manifest.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8314-466d-ad02-f94d64681643 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8314-466d-ad02-f94d64681643 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8314-466d-ad02-f94d64681643 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8314-466d-ad02-f94d64681643 | FLAKY | 1/2 |
| rptest.tests.node_pool_migration_test.NodePoolMigrationTest.test_migrating_redpanda_nodes_to_new_pool.balancing_mode=node_add.test_mode=TestMode.FAST_MOVES.cleanup_policy=compact.delete | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8314-466d-ad02-f94d64681643 | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61892#01950dd8-8311-49ba-a74e-4a5b41d82b54 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61895">build#61895</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.mount_unmount_test.MountUnmountIcebergTest.test_simple_unmount.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61895#01951001-07ac-4bd1-a3a5-3faf5d47472b | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.EndToEndHydrationTimeoutTest.test_hydration_completes_when_consumer_killed | ducktape | https://buildkite.com/redpanda/redpanda/builds/61895#01951015-6524-4324-8dd2-4c332093274e | FLAKY | 1/2 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=False.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61895#01951001-07ac-4bd1-a3a5-3faf5d47472b | FAIL | 0/1 |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=False.with_iceberg=False.with_chunked_compaction=True.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61895#01951001-07ac-4bd1-a3a5-3faf5d47472b | FLAKY | 1/3 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/61895#01951015-6522-4df0-b5ed-3d2f0dfdfc81 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61897">build#61897</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61897#019510b0-883a-4b05-8a7b-6267372a9fbb | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61897#019510f3-fa45-4527-9c05-dcbe295159d4 | FLAKY | 1/2 |
| rptest.tests.datalake.coordinator_retention_test.CoordinatorRetentionTest.test_retention.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61897#019510f3-fa46-4854-8400-dfb026937523 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61897#019510f8-5a58-4b89-97ea-4b5d5c04faa2 | FLAKY | 1/3 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61897#019510f8-5a59-4e74-acf3-2e704bac19a0 | FLAKY | 1/2 |
| rptest.tests.read_distribution_test.ReadDistributionTest.test_read_distribution_metric | ducktape | https://buildkite.com/redpanda/redpanda/builds/61897#019510f3-fa45-4527-9c05-dcbe295159d4 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61935">build#61935</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61935#01951642-1ec0-4d40-b9e9-a753975d6407 | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61935#01951656-b23c-40a9-9dd9-92c263c77c68 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=10 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61935#01951656-b23c-40a9-9dd9-92c263c77c68 | FLAKY | 1/2 |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61935#01951656-b23c-40a9-9dd9-92c263c77c68 | FLAKY | 1/3 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61965">build#61965</a></summary>


| test_id                                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                           | unit        | https://buildkite.com/redpanda/redpanda/builds/61965#019519c3-db51-4d0e-8158-905cb4b58da5 | FLAKY         | 1/2      |
| rptest.tests.cloud_storage_chunk_read_path_test.CloudStorageChunkReadTest.test_read_chunks                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a0e-7acc-4578-a838-c6f6f8e5add3 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a0e-7acd-41fd-a3b9-09d0d156a6b2 | FLAKY         | 1/8      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a22-31f1-4bca-bb12-2b91322fb6d9 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a22-31ee-4189-8dc9-d4de4512d31b | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.delete.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a22-31ef-4abf-827b-193b5dd92585 | FLAKY         | 1/2      |
| rptest.tests.log_compaction_test.LogCompactionTest.compaction_stress_test.cleanup_policy=compact.key_set_cardinality=1000.storage_compaction_key_map_memory_kb=3        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61965#01951a22-31f1-4bca-bb12-2b91322fb6d9 | FLAKY         | 1/2      |
</details>
<!--/body-->

nvartolomei commented:
What's the current behavior when we stumble on a snappy batch? How did you find this difference?

Can you add a "conformance/e2e" test by writing a batch with java's snappy implementation, re-compressing it with rp, reading it back?

WillemKauf committed:
`tests`: add `python-snappy` and `lz4` as `rptest` dependencies

In order to allow `kafka-python` to use these compression types,
we must be able to import the respective module.

WillemKauf committed:
`rptest`: add `java_compression_test.py`

To test compression compatibility with Java-based Kafka consumers/producers.

These tests are parameterized for all compression types, but they most notably
serve as reproducers for an outstanding header-field encoding bug in
`snappy_java_compressor.cc`.

WillemKauf commented:
Push to:
* Add `java_compression_test.py` to test end to end `snappy` compression via a `VerifiableProducer`, decompression/recompression of the batches with `redpanda`, and finally consuming via a `KafkaConsumer` (from `kafka-python`). This test will fail with a `snappy.snappy.UncompressError` without the changes made in this PR from little-endian format to big-endian format.

WillemKauf commented:
>How did you find this difference?

An issue was raised yesterday, see https://github.com/redpanda-data/redpanda/issues/25091.

>What's the current behavior when we stumble on a snappy batch?

We encode the `version` and `compatible_version` fields via little endian, ending up with the value `16777216` instead of `1` when decoded via a `snappy` client. According to the above issue, most `snappy` clients don't actually check these fields, but `kafka-python` does, leading to a `snappy.snappy.UncompressError`. That is why this has gone undiscovered for several years (the fields will decode for most clients, but to an incorrect value).

> Can you add a "conformance/e2e" test by writing a batch with java's snappy implementation, re-compressing it with rp, reading it back?

Added a ducktape test!

**Concerning backwards compatibility with existing compressed batches**, we do [have a check in `redpanda`](https://github.com/redpanda-data/redpanda/blob/59ee571fe90251804f1e53afd1857b9466f5df10/src/v/compression/internal/snappy_java_compressor.cc#L89) around the decoded `min_compatible_version`, but decoding a previously little-endian encoded value using big-endian should still result in a `value > min_compatible_version`.

nvartolomei commented:
The change looks good. I also tried to work through the scenario where we have already wrote data to disk/cloud with incorrectly encoded header and it seems to me it will work ok.

- Can you please add a note in the decompressor (maybe in the snappy_magic struct too) that the compressor needs to remain compatible with these wrongly encoded headers. I.e. it must treat value `16777216` as `1`.

- Please add a backwards compatibility (unit) test too? Compress some payload (i.e. commit it as a file to the repo) using old version and decompress it with current version.

- And add a comment about the mechanics of the test. E.g. "produce snappy compressed data with java client to exercise the java flavour of snappy serialization (distinct magic header), wait for compaction so that we get redpanda to re-compress the data, read with the python reader  (one of the few clients that actually validate the header) \[link to the issue\])

WillemKauf committed:
`compression`: add backwards compatibility check for `snappy`

The two committed files in `snappy_payload` are a raw uncompressed data file,
and a `snappy` compressed data file generated by `redpanda` using the
incorrect little-endian encoding for the version fields in the `snappy`
header.

They are used in a unit test to ensure that with the big-endian fix for
`snappy`, we are still able to decompress the buffer and get the same
decompressed data as before the fix.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61895-->
<!--41393e6c-c8c5-4c7a-a6fc-171548d4d3d1-->
<!--title-->
## Retry command for [Build#61895](https://buildkite.com/redpanda/redpanda/builds/61895)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"enable_failures":true,"mixed_versions":false,"with_chunked_compaction":true,"with_iceberg":false,"with_tiered_storage":true}
```
<!--/body-->

WillemKauf commented:
Force push to:
* Add comments describing need for backwards compatibility throughout `snappy` related implementation files
* Add unit test for backwards `snappy` compatibility post big-endian fix
* Add a new ducktape test, `test_upgrade_java_snappy_compression`, to ensure backwards compatibility after a `redpanda` upgrade with a mixture of little- and big-endian `snappy` encoded batches.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61897-->
<!--a07d9add-cdf2-4f95-a140-ab17b2e11542-->
<!--title-->
## Retry command for [Build#61897](https://buildkite.com/redpanda/redpanda/builds/61897)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/read_distribution_test.py::ReadDistributionTest.test_read_distribution_metric
```
<!--/body-->

WillemKauf commented on src/v/compression/tests/snappy_tests.cc:
```diff
@@ -137,3 +139,32 @@ TEST(SnappyTest, CompressedVersionHeadersSnappyJavaTest) {
+    // the existing, improperly encoded batches (as version, min_version fields
+    // with value 1 will decode to the value 16777216).
+    // See: https://github.com/redpanda-data/redpanda/issues/25091
+    auto root = std::filesystem::path(
```
What is the best way to make this path CI friendly? :)

rockwotj replied:
Honestly to inline them in C++ would be easiest.

Otherwise there are existing patterns in the crypto and wasm tests to deal with passing files into tests from the filesystem. Probably using an environment variable is easiest.

rockwotj commented on src/v/compression/internal/snappy_java_compressor.cc:
```diff
@@ -83,9 +76,14 @@ iobuf snappy_java_compressor::uncompress(const iobuf& x) {
+    // with value 1 will decode to the value 16777216).
+    const auto version = iter.consume_be_type<int32_t>();
+    const auto min_version = iter.consume_be_type<int32_t>();
     if (unlikely(min_version < snappy_magic::min_compatible_version)) {
```
How does this check not fire before this PR? Do we never decode Snappy values (in tests or in real life)

nvartolomei replied:
Well as long as min_version >= 1 this check will always pass. I'm pretty sure this check had a good intention but it was implemented incorrectly. The only check that makes sense to me here is `if (snappy_magic::version < min_version) throw ...;`, i.e. allow encoders to require minimum version of a decoder. Probably too late for that...

FWIW, this code is also incorrectly translated from Java. The java check is `version < snappy_magic::min_compatible_version`. min_version is not used by java anywhere directly: https://github.com/search?q=repo%3Axerial/snappy-java%20compatibleVersion&type=code

rockwotj commented on tests/rptest/tests/compatibility/java_compression_test.py:
```diff
@@ -0,0 +1,166 @@
+        self.topic_spec = TopicSpec(
+            name="tapioca",
+            replication_factor=1,
+            compression_type=TopicSpec.COMPRESSION_SNAPPY,
```
This does nothing FWIW

WillemKauf replied:
Removed.

rockwotj commented on tests/rptest/tests/compatibility/java_compression_test.py:
```diff
@@ -0,0 +1,166 @@
+        # (little-endian and big-endian)
+        expected_num_segments = 8
+        self.produce_until_segment_count(expected_num_segments,
+                                         compression_type="snappy")
```
While we are here,do we have tests for zstd and lz4? Can we add all the codecs to this test? Seems wild if we don't have them already.

WillemKauf replied:
Done!

WillemKauf committed:
`compression`: remove version check for `snappy_java_compressor`

Most `snappy` clients do not perform this version check, and furthermore,
it is implemented incorrectly here.

WillemKauf commented:
Force push to:
* Make required data files for `LittleEndianHeadersBackwardsCompatibilitySnappyJavaTest` dependencies in `CMakeLists.txt`/`BUILD` for portability
* Remove version check in `snappy_java_compressor.cc`.
* Parameterize `test_java_compression` and `test_upgrade_java_compression` with all compression types.
* Add `lz4` as a Python ducktape dependency.

rockwotj reviewed (dismissed):
LGTM - thanks for the comprehensive fix and comments/commits!

rockwotj commented:
@WillemKauf should we consider backporting since this is a bug fix?

WillemKauf commented:
@rockwotj I think a backport makes sense. I think I'll have to amend the `RedpandaVersionTriple` used in the backwards compatibility test to a more specific version of `24.3` to match test expectations, but I'll add the backports for `24.3` and `24.2`.

rockwotj commented:
I am also okay if we drop that test for the backports

WillemKauf commented:
> I am also okay if we drop that test for the backports

SG

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61921-->
<!--1e93cdae-eb0b-4b25-9c5c-ba2f4ea9ad04-->
<!--title-->
## Retry command for [Build#61921](https://buildkite.com/redpanda/redpanda/builds/61921)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/read_distribution_test.py::ReadDistributionTest.test_read_distribution_metric
```
<!--/body-->

WillemKauf commented:
Force push to:
* Fix improper cleanup of `VerifiableProducer` in `java_compression_test.py`, which would contaminate other ducktape test environments and cause them to fail.

rockwotj commented on tests/rptest/tests/compatibility/java_compression_test.py:
```diff
@@ -0,0 +1,181 @@
+            lambda: self.partition_segments() >= count,
+            timeout_sec=timeout_sec,
+            err_msg=f"Timed out waiting for {count} segments to be produced.")
+        producer.stop()
+        producer.clean()
+        producer.free()
```
If the `wait_until` fails then we'll leak these right? Can we use VerifiableProducer as a context manager or add a `try: ... finally: ...`?

WillemKauf replied:
Added `try`/`finally` block here.

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25092-v24.2.x-207 remotes/upstream/v24.2.x
git cherry-pick -x 035220123a 1c1b00601c 5723eb4bc4 72d02ee7ba a84252d4b0 17a2e550d4 379f380d19
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13403429479).

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25092-v24.3.x-649 remotes/upstream/v24.3.x
git cherry-pick -x 035220123a 1c1b00601c 5723eb4bc4 72d02ee7ba a84252d4b0 17a2e550d4 379f380d19
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13403429638).

dotnwat reviewed (commented):
did we not consider writing big endian, and then detecting versions previously written with little endian and converting them correctly? i'm assuming we could reasonably detect the little endian versions by checking some bits?

rockwotj commented:
We do this in compaction with these changes, we don't alter anything in flight in the broker path, because we don't know if the batch came from us or not, and the perf impact is unclear. There was a thread about this in Slack, but I don't see the benefit since this bug has been around forever and only just detected, all the clients I looked at don't check these version numbers at all

WillemKauf commented:
/backport v24.1.x

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-25092-v24.1.x-345 remotes/upstream/v24.1.x
git cherry-pick -x 035220123a 1c1b00601c 5723eb4bc4 72d02ee7ba a84252d4b0 17a2e550d4 379f380d19
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/13467529316).
